db/hints/manager.cc

```diff
@@ -1004,6 +1017,73 @@ db::replay_position manager::end_point_hints_manager::sender::send_one_file_ctx:
     return rp;
 }
 
+manager::end_point_hints_manager::sender::concurrency_limiter::concurrency_limiter(size_t max_limit)
+        : _max_limit(max_limit)
+        , _next_update_time(clock::now())
+{}
+
+future<semaphore_units<>> manager::end_point_hints_manager::sender::concurrency_limiter::get_send_units_for(size_t mutation_size) {
+    // Limit the size of hints being concurrently sent so that it is lower than _current_limit.
+    // If a mutation larger than the limit is encountered, allow sending it
+    // but not concurrently with other hints.
+    auto f = get_units(_sem, std::min(_current_limit, mutation_size));
+
+    // If the semaphore has insufficient units, we mark a flag in order to increase
+    // the number of units later.
+    _need_increase |= (mutation_size > _current_limit) || (_sem.waiters() > 0);
+
+    return f;
+}
+
+manager::end_point_hints_manager::sender::clock::duration manager::end_point_hints_manager::sender::concurrency_limiter::get_delay_before_next_send() {
+    if (_current_limit > 1) {
+        return std::chrono::seconds(_next_delay_distribution.min());
+    } else {
+        return std::chrono::seconds(_next_delay_distribution(rand_gen));
+    }
+}
+
+void manager::end_point_hints_manager::sender::concurrency_limiter::account_successful_write(size_t size) {
+    _size_written_since_last_update += size;
+    _count_written_since_last_update++;
+
+    const auto now = clock::now();
+    if (!_need_increase || now < _next_update_time) {
+        return;
+    }
+
+    // Increase the concurrency semaphore by 10% of the mean mutation size
+    auto increase = _size_written_since_last_update / (10 * _count_written_since_last_update);
+    increase = std::max(increase, size_t(1));
+
+    const auto new_limit = std::min(_current_limit + increase, _max_limit);
```

Comment:
nitpick: in case of `size_t`, I think it's better to explicitly state the type instead of typing `auto` - it's not much more typing and it makes it clear what's the expected type. This comment applies to a couple of similar usages below as well.